content/posts/networkx/vf2pp/graph-iso-vf2pp/index.md

```diff
@@ -0,0 +1,132 @@
+
+## Introduction
+
+We should quickly remind at this point, that the individual functionalities already implemented are:
+
+- **Node ordering** which finds the optimal order to access the nodes, such that those that are more likely to match, are placed first in the order. This reduces the possibility of infeasible searches taking place first.
+- **Candidate selection** such that, given a node $u$ from $G_1$, we obtain the candidate nodes $v$ from $G_2$.
+- **Feasibility rules** introducing easy-to-check cutting and consistency conditions which, if satisfied by a candidate pair of nodes $u$ from $G_1$ and $v$ from $G_2$, the mapping is extended.
+- **$T_i$ updating** which updates the $T_i$ and $\tilde{T}_i$, $i=1,2$ parameters in case that a new pair is added to the mapping, and restores them when a pair is popped from it.
+
+We are going to use all these functionalities to form our **Isomorphism solver**.
+
+## VF2++
+
+First of all, let's describe the algorithm in simple terms, before presenting the pseudocode. The algorithm will look something like this:
+
+1. Check if all **preconditions** are satisfied before calling the actual solver. For example there's no point examining two graphs with different number of nodes for isomorphism.
+2. Initialize all the necessary **parameters** ($T_i$, $\tilde{T}_i$, $i=1,2$) and maybe cache some information that is going to be used later.
+3. Take the next unexamined node $u$ from the ordering.
+4. Find its candidates and check if there's a candidate $v$ such that the pair $u-v$ satisfies the **feasibility rules**
+5. if there's any, extend the mapping and **go to 3**.
+6. if not, pop the last pair $\hat{u}-\hat{v}$ from the mapping and try a different candidate $\hat{v}$, from the remaining candidates of $\hat{u}$
+7. The two graphs are **isomorphic** if the number of **mapped nodes** equals the number of nodes of the two graphs.
+8. The two graphs are **not isomorphic** if there are no remaining candidates for the first node of the ordering (root).
+
+The pseudocode for the **VF2++** is presented below.
+
+> ---
+>
+> **VF2++**
+>
+> ---
+>
+> 1. **Set** Check if the number of nodes and degree sequence is the same for both graphs.
+> 2. **Set** $T_1, T_2, \tilde{T}_1, \tilde{T}_2$ to default
+> 3. **Set** $m = m^{-1} = \{\}$
+> 4. $order=\def\textsc#1{\dosc#1\csod}{matchingOrder()}$
+> 5. $root=order\[0\]$
+> 6. **Set** $stack=\[\(root, \def\textsc#1{\dosc#1\csod}{findCandidates(root)}\)\]$
+> 7. **while** $stack$ **do**
+>    - $n,nc=stack.peek()$
+>    - **if** $nc$ is empty **do**
+>      - $stack.pop()$
+>      - $\def\textsc#1{\dosc#1\csod}{restoreState()}$
+>      - $continue$
+>    - **if** $\def\textsc#1{\dosc#1\csod}{feasibility(n, nc.pop())}$ **do**
+>      - **if** len(m) equals len(G1) **return True**
+>      - $\def\textsc#1{\dosc#1\csod}{updateState()}$
+>      - $nextNode=order.next()$
+>      - $stack.append\(\(nextNode, \def\textsc#1{\dosc#1\csod}{findCandidates(nextNode)}\)\)$
```

Comment:
One of the great things about Python is it's readability - in principle it'd be great if, instead of "rewriting" the algorithm in pseudocode, we could instead show the actual implementation!

For example, [the actual implementation](https://github.com/networkx/networkx/blob/bffcd74649fb95a57fb834846eb3c7d9693c55b8/networkx/algorithms/isomorphism/vf2pp.py#L171-L236) is only slightly longer, and IMO is quite readable. I'm not sure whether there's a nice way to embed code directly from github, or whether you'd have to copy it into a code-block, but I think that'd be a nice improvement.